### PR TITLE
Implemented logic for storing and applying multiple field mappings(#2hran7u)

### LINKED ISSUE #81: Functionality to save multiple field mappings
## What are the acceptance criteria?

- [ ] ...

## Can you complete this feature request by yourself?

- [ ] YES
- [ ] NO

### PR DESCRIPTION
**Pull request by disha1202**
### Related Issues

Closes #81 

### Screenshots of Visual Changes before/after (If There Are Any)

[screen-recorder-fri-nov-04-2022-12-57-09.webm](https://user-images.githubusercontent.com/58461801/199936863-6afb5082-366e-4963-b269-da6e67561251.webm)

Screenshot from 2022-11-04 15-03-15

**IMPORTANT NOTICE** - Remember to add changelog entry

### Contribution and Currently Important Rules Acceptance

- [ ] I read and followed [contribution rules](https://github.com/hotwax/import#contribution-guideline)

### CODE REVIEW

File: src/views/PurchaseOrder.vue [hunk 1]
```diff
@@ -17,55 +17,70 @@
...
 
           <ion-item>
             <ion-label>{{ $t("Facility ID") }}</ion-label>
-            <ion-select interface="popover" v-if="content.length" :placeholder = "$t('Select')" v-model="facilityField">
+            <ion-select interface="popover" v-if="content.length" :placeholder = "$t('Select')" v-model="fields.facility">
               <ion-select-option v-bind:key="index" v-for="(prop, index) in Object.keys(content[0])">{{ prop }}</ion-select-option>
             </ion-select>
           </ion-item>
-        </ion-list>
+        </ion-list>
```
Comment (ymaheshwari1) [inline src/views/PurchaseOrder.vue:62]: Unwanted change.

File: src/views/PurchaseOrder.vue [hunk 2]
```diff
@@ -17,55 +17,70 @@
 
         <ion-list>
           <ion-list-header>{{ $t("Select the column index for the following information in the uploaded CSV.") }}</ion-list-header>
+
+          <ion-item lines="none">
+            <ion-label>{{ $t("Select mapping") }}</ion-label>
+            <ion-select interface="popover" @ionChange="mapFields">
+              <ion-select-option v-for="mapping in fieldMappings" :value="mapping" :key="mapping">{{ mapping.name }}</ion-select-option>
```
Comment (ymaheshwari1) [inline src/views/PurchaseOrder.vue]: Use a unique value as a key instead of using an object.

File: src/views/PurchaseOrder.vue [hunk 3]
```diff
@@ -13,59 +13,74 @@
           <ion-label>{{ file.name ? $t("Purchase order ") +  file.name : $t('Purchase order') }}</ion-label>
           <input @change="getFile" ref="file" class="ion-hide" type="file" id="inputFile"/>
           <label for="inputFile">{{ $t("Upload") }}</label>
-        </ion-item>      
+        </ion-item> 
+
+        <ion-item lines="none">
+          <ion-label>{{ $t("Select mapping") }}</ion-label>
+          <ion-select interface="popover" @ionChange="mapFields">
+            <ion-select-option v-for="mapping in fieldMappings" :value="mapping" :key="mapping">{{ mapping.name }}</ion-select-option>
```
Comment (ymaheshwari1) [inline src/views/PurchaseOrder.vue:21]: What if there are no mappings available?

File: src/views/PurchaseOrder.vue [hunk 4]
```diff
@@ -13,59 +13,74 @@
...
+        <ion-button color="medium" @click="review" expand="block">
           {{ $t("Review") }}
           <ion-icon slot="end" :icon="arrowForwardOutline" />
         </ion-button>
+
+        <ion-item>
+          <ion-label>{{ $t("Mapping name") }}</ion-label>
+          <ion-input v-model="mappingName" />
+        </ion-item>
+        <ion-button @click="savaMapping">{{ $t("Save mapping") }}</ion-button>
```
Comment (ymaheshwari1) [inline src/views/PurchaseOrder.vue:73]: I think we should not allow user to save mapping without a mapping name.

**Review by ymaheshwari1 — CHANGES_REQUESTED** [review]

File: src/views/PurchaseOrder.vue (hunk 1 above)
Comment (disha1202) [reply src/views/PurchaseOrder.vue]: Reverted

File: src/views/PurchaseOrder.vue (hunk 2 above)
Comment (disha1202) [reply src/views/PurchaseOrder.vue]: Made the requested change

File: src/views/PurchaseOrder.vue (hunk 3 above)
Comment (disha1202) [reply src/views/PurchaseOrder.vue]: Disabled select option.

File: src/views/PurchaseOrder.vue (hunk 4 above)
Comment (disha1202) [reply src/views/PurchaseOrder.vue:73]: Made the requested change

File: src/views/PurchaseOrder.vue [hunk 5]
```diff
@@ -92,15 +108,30 @@ export default defineComponent({
...
         orderItemsList: [],
       }
     },
+    computed: {
+      ...mapGetters({
+        fieldMappings: 'user/getFieldMappings'
+      })
+    },
     methods: {
+      savaMapping() {
```
Comment (ymaheshwari1) [inline src/views/PurchaseOrder.vue:128]:
```suggestion
      saveMapping() {
```

**Review by ymaheshwari1 — CHANGES_REQUESTED** [review]

**Review by ymaheshwari1 — APPROVED** [review]

File: src/store/modules/user/mutations.ts [hunk 6]
```diff
@@ -19,6 +19,14 @@ const mutations: MutationTree <UserState> = {
     },
     [types.USER_INSTANCE_URL_UPDATED] (state, payload) {
         state.instanceUrl = payload;
+    },
+    [types.USER_FIELD_MAPPINGS_UPDATED] (state, payload) {
+        const mapping = state.fieldMappings.find((mapping: any) => mapping.name === payload.name );
```
Comment (adityasharma7) [inline src/store/modules/user/mutations.ts:24]: I think this should be part of action

File: src/views/PurchaseOrder.vue [hunk 7]
```diff
@@ -92,15 +108,30 @@ export default defineComponent({
...
         content: [],
-        orderIdField: "",
-        productSkuField: "",
-        dateField: "",
-        quantityField: "",
-        facilityField: "",
+        fields: {
+          orderId: "",
+          productSku: "",
+          date: "",
```
Comment (adityasharma7) [inline src/views/PurchaseOrder.vue:114]:
```suggestion
          orderDate: "",
```

File: src/store/modules/user/index.ts [hunk 8]
```diff
@@ -12,6 +12,7 @@ const userModule: Module<UserState, RootState> = {
       current: null,
       currentFacility: {},
       instanceUrl: '',
+      fieldMappings: []
```
Comment (adityasharma7) [inline src/store/modules/user/index.ts:15]: We could have it as object with key as mappingPrefId and value as object having mappingPrefId, mappingPrefName, mappingPrefTypeEnumId and mappingPrefValue

File: src/views/PurchaseOrder.vue [hunk 9]
```diff
@@ -92,15 +108,30 @@ export default defineComponent({
       return {
         file: "",
         content: [],
-        orderIdField: "",
-        productSkuField: "",
-        dateField: "",
-        quantityField: "",
-        facilityField: "",
+        fields: {
```
Comment (adityasharma7) [inline src/views/PurchaseOrder.vue:111]: Why not fieldMapping?

File: src/views/PurchaseOrder.vue [hunk 10]
```diff
@@ -112,28 +143,35 @@ export default defineComponent({
...
+          name:'PurchaseOrderDetail'
+        })
+      },
+      mapFields(event) {
+        if(event && event.detail.value){
+          const mapping = JSON.parse(JSON.stringify(event.detail.value));
+          this.fields = mapping.fieldMapping;
+          this.mappingName = mapping.name;
+        }
         this.orderItemsList = this.content.map(item => {
```
Comment (adityasharma7) [inline src/views/PurchaseOrder.vue:164]: I think this should be done with review only

File: src/views/PurchaseOrder.vue [hunk 11]
```diff
@@ -112,28 +143,35 @@ export default defineComponent({
...
+        if(!this.orderItemsList.length) this.mapFields();
+        this.store.dispatch('order/updatedOrderList', this.orderItemsList);
+        this.router.push({
+          name:'PurchaseOrderDetail'
+        })
+      },
+      mapFields(event) {
+        if(event && event.detail.value){
+          const mapping = JSON.parse(JSON.stringify(event.detail.value));
+          this.fields = mapping.fieldMapping;
```
Comment (adityasharma7) [inline src/views/PurchaseOrder.vue:161]:
```suggestion
          this.fields = mapping.fieldMapping;
```

File: src/views/PurchaseOrder.vue [hunk 12]
```diff
@@ -112,28 +143,35 @@ export default defineComponent({
...
       },
-      mapFields() {
+      review() {
+        if(!this.orderItemsList.length) this.mapFields();
+        this.store.dispatch('order/updatedOrderList', this.orderItemsList);
+        this.router.push({
+          name:'PurchaseOrderDetail'
+        })
+      },
+      mapFields(event) {
```
Comment (adityasharma7) [inline src/views/PurchaseOrder.vue:158]: We could check if mapping fields exist in csv

**Review by adityasharma7 — CHANGES_REQUESTED** [review]

**Comment by adityasharma7** [comment]
@disha1202 

Could you please share some screenshots and videos for the change?

**Comment by disha1202** [comment]
> @disha1202
> 
> Could you please share some screenshots and videos for the change?

Sure Sir, added in the PR description.

File: src/views/PurchaseOrder.vue (hunk 12 above)
Comment (disha1202) [reply src/views/PurchaseOrder.vue:168]: Made the requested change.
Also displayed a toast if any of the fields is missing.

File: src/views/PurchaseOrder.vue (hunk 10 above)
Comment (disha1202) [reply src/views/PurchaseOrder.vue]: Made the requested change.

File: src/views/PurchaseOrder.vue (hunk 9 above)
Comment (disha1202) [reply src/views/PurchaseOrder.vue]: Updated the variable name

File: src/store/modules/user/index.ts (hunk 8 above)
Comment (disha1202) [reply src/store/modules/user/index.ts]: Created it as an object with key as mapping name and value as mapping fields.

File: src/store/modules/user/mutations.ts (hunk 6 above)
Comment (disha1202) [reply src/store/modules/user/mutations.ts]: Updated the code.

File: src/views/PurchaseOrder.vue [hunk 13]
```diff
@@ -92,21 +107,42 @@ export default defineComponent({
...
+          orderDate: "",
+          quantity: "",
+          facility: "",
+        },
+        mappingName: "",
         orderItemsList: [],
       }
     },
     methods: {
+      generateUniqueId(id) {
```
Comment (adityasharma7) [inline src/views/PurchaseOrder.vue:132]: Improve method name, this is not generic method to generate id for any use case

**Review by adityasharma7 — CHANGES_REQUESTED** [review]

File: src/views/PurchaseOrder.vue [hunk 14]
```diff
@@ -120,30 +150,42 @@ export default defineComponent({
...
           })
         } else {
           showToast(translate("Select all the fields to continue"));
+        } 
+      },
+      mapFields(event) {
+        if(event && event.detail.value) {
+          const fieldMapping = JSON.parse(JSON.stringify(event.detail.value));
+          for(const field in fieldMapping.mappingPrefValue){
+            if(!Object.keys(this.content[0]).includes(fieldMapping.mappingPrefValue[field])) fieldMapping.mappingPrefValue[field] = "";
```
Comment (adityasharma7) [inline src/views/PurchaseOrder.vue:185]: We could store Object.keys(this.content[0]) to reuse

File: src/views/PurchaseOrder.vue [hunk 15]
```diff
@@ -120,30 +150,42 @@ export default defineComponent({
...
             name:'PurchaseOrderDetail'
           })
         } else {
           showToast(translate("Select all the fields to continue"));
+        } 
+      },
+      mapFields(event) {
+        if(event && event.detail.value) {
+          const fieldMapping = JSON.parse(JSON.stringify(event.detail.value));
+          for(const field in fieldMapping.mappingPrefValue){
```
Comment (adityasharma7) [inline src/views/PurchaseOrder.vue:184]: We could use reduce() here

**Review by adityasharma7 — CHANGES_REQUESTED** [review]

File: src/views/PurchaseOrder.vue [hunk 16]
```diff
@@ -119,30 +150,51 @@ export default defineComponent({
...
+      },
+      mapFields(event) {
+        if(event && event.detail.value) {
+          const fieldMapping = JSON.parse(JSON.stringify(event.detail.value));
+          const CsvFields = Object.keys(this.content[0]);
+
+          const missingFields = Object.values(fieldMapping.mappingPrefValue).filter(field => {
+            if(!Object.keys(this.content[0]).includes(field)) return field;
+          });
+          if(missingFields.length) showToast(translate(`Some of the mapping fields are missing in the CSV: \n ${ missingFields.join(", ") }`))
```
Comment (adityasharma7) [inline src/views/PurchaseOrder.vue:189]: I think we could pass missingFields as variable
Refer this code
https://github.com/hotwax/preorder/blob/main/src/views/product-details.vue#L347

**Review by adityasharma7 — CHANGES_REQUESTED** [review]

**Pull request merged**